### DISCUSSION TITLE
CSLA 6 any way to have strongly typed criteria parameters?

### DISCUSSION THREAD

**Discussion by michaelcsikos** [MarimerLLC/csla#2825] [Questions]
Previously with static factory methods it was easy to ensure we were passing the correct parameters for the `DataPortal_XYZ` methods. Now with `IDataPortal<T>` we're dealing with `params object[] criteria` so it's easy to make a mistake or introduce a bug. Tests help, of course, but a compile error is better. Is a `[Serializable]` criteria class the simplest solution? ~~Is there any advantage gained by using `CriteriaBase<T>`?~~ It seems `CriteriaBase<T>` is required.

**Comment by rockfordlhotka** [comment]
I hear what you are saying. 

Using a strongly-typed "DTO" based on `MobileObject` or `CriteriaBase` seems like a reasonable way to address the issue.

I'm open to other suggestions as well if anyone has ideas.

**Reply by michaelcsikos** [reply]
Thanks @rockfordlhotka. Thinking about this some more, is there any reason we can't create `static` methods to help enforce the correct parameters?

```cs
public static async Task<Person> NewPersonAsync(Guid companyId, bool isUser, IDataPortal<Person> portal)
{
    return await portal.CreateAsync(companyId, isUser);
}

public static async Task<Person> GetPersonAsync(Guid id, IDataPortal<Person> portal)
{
    return await portal.FetchAsync(id);
}
```

Or they could be extension methods:

```cs
public static async Task<Person> NewPersonAsync(this IDataPortal<Person> portal, Guid companyId, bool isUser)
{
    return await portal.CreateAsync(companyId, isUser);
}
```

**Reply by rockfordlhotka** [reply]
You can create static methods. There is at least one existing discussion thread on that topic, and to make it easier we added an IDataPortalFactory that you can inject and pass to the static method, so inside the static method you can access multiple data portal types if necessary.

**Reply by rockfordlhotka** [reply]
You can also create concrete factory types, which (I think) is my current favorite way to resolve this issue.

```c#
public class PersonFactory
{
  public PersonFactory(IDataPortal<Person> personPortal)
  {
    PersonPortal = personPortal;
  }

  private IDataPortal<Person> PersonPortal { get; set; }

  public async Task<Person> GetPersonAsync(Guid id)
  {
    return await PersonPortal.FetchAsync(id);
  }
}
```

**Reply by michaelcsikos** [reply]
I like this approach. As a nested class where we used to put the `#region Factory methods` it's a more familiar paradigm:

```cs
#region Factory methods

public class Factory
{
    IDataPortal<Person> DataPortal { get; set; }

    public Factory(IDataPortal<Person> dataPortal)
    {
        DataPortal = dataPortal;
    }

    public Person New(Guid companyId, bool isUser)
    {
        return DataPortal.Create(companyId, isUser);
    }

    public Person Get(Guid id)
    {
        return DataPortal.Fetch(id);
    }

    public async Task<Person> GetAsync(Guid id)
    {
        return await DataPortal.FetchAsync(id);
    }

    public void Delete(Guid id)
    {
        DataPortal.Delete(id);
    }

    public async Task DeleteAsync(Guid id)
    {
        await DataPortal.DeleteAsync(id);
    }
}

#endregion
```
